queue/driver_mongo.go

```diff
@@ -1400,14 +1268,146 @@ type dispatchAttemptInfo struct {
 	// misses is the number of times the dispatcher attempted to dispatch a job
 	// but failed.
 	misses int
+	// total is the total number of documents the cursor returned.
+	total int
+}
+
+// dispatchStaleAndPending attempts to dispatch a job. Stale in progress jobs are given precedence over pending jobs.
+func (d *mongoDriver) dispatchStaleAndPending(ctx context.Context, startAt time.Time) (amboy.Job, dispatchAttemptInfo, error) {
+	now := time.Now()
+	var info dispatchAttemptInfo
+
+	job, dispatchInfo, err := d.tryDispatchJob(ctx, d.getNextStaleInProgressQuery(now), d.opts.SampleSize, startAt)
+	info.misses += dispatchInfo.misses
+	info.skips += dispatchInfo.skips
+	if err != nil {
+		return nil, info, errors.Wrap(err, "dispatching stale in progress job")
+	}
+
+	if job != nil || d.opts.SampleSize > 0 && dispatchInfo.total >= d.opts.SampleSize {
```

Comment:
It seems like it would be preferable to continue trying to dispatch pending jobs even if it already iterated `SampleSize` # of in progress jobs to avoid overly blocking the pending queue on the stale in progress jobs.

Reply:
done